samples/nrf9160/modem_shell/src/cloud/cloud_shell.c

```diff
@@ -184,17 +190,24 @@ static void nrf_cloud_event_handler(const struct nrf_cloud_evt *evt)
 			/* data was valid; no need to pass to other handlers */
 			break;
 		} else if (err == -EFAULT) {
-			/* data was an A-GPS error; no need to pass to other handlers */
+			/* data were an A-GPS error; no need to pass to other handlers */
 			break;
 		}
 #endif
 #if defined(CONFIG_NRF_CLOUD_PGPS)
-		err = nrf_cloud_pgps_process(evt->data.ptr, evt->data.len);
-		if (err) {
-			mosh_error("Error processing P-GPS packet: %d", err);
+		/* Check if we are waiting for P-GPS data in order to avoid parsing other data as
+		 * P-GPS data and showing irrelevant error messages.
+		 */
+		if (nrf_cloud_pgps_loading()) {
+			err = nrf_cloud_pgps_process(evt->data.ptr, evt->data.len);
+			if (err) {
+				mosh_error("Error processing P-GPS packet: %d", err);
+			}
```

Comment:
This does not handle the case that a JSON message was received which is not a MoSh command, and not an A-GPS or P-GPS error, the same way as you handle such a message when P-GPS is not enabled.  Down on line 208, you check the err and either print nothing or print an error.

Reply:
Hopefully I understood correctly what you are saying. Please check my latest modifications. Now the handling should be similar in all cases except when P-GPS is enabled and loading is ongoing. In such case whatever msg that is not A-GPS data, A-GPS error or MoSh command will be fed into the P-GPS processor and will trigger an error even if it has nothing to do with P-GPS. But I can live with that. And all this code will eventually be removed when we get separate topics for A-GPS, P-GPS and whatnot.

Reply:
It's almost correct now.  Please take a look at how ATv2 does this (in `applications/asset_tracker_v2/src/modules/cloud_module.c function `agps_data_handle()`).  The call to `nrf_cloud_pgps_loading()` should be removed here.  The function `nrf_cloud_pgps_process()` is what sets the state to `PGPS_LOADING` which `nrf_cloud_pgps_loading()` tests for, so your code as written will ignore the P-GPS data.

Reply:
I am calling nrf_cloud_pgps_loading() to avoid feeding non-PGPS data to P-GPS process function and by doing so avoid showing parsing errors. nrf_cloud_pgps_loading() also returns true in case state == PGPS_REQUESTING which seems to be the case in my test so there is no issue here as far as I can tell.

In ATv2 the return value of nrf_cloud_pgps_notify_prediction() seems to determine if nrf_cloud_pgps_process() gets called or not (by returning early in some cases) whereas I just put this function into the queue. To be honest, I don't understand the whole function nrf_cloud_pgps_notify_prediction() very well - why is it called upon receiving A-GPS data from cloud in the first place?

That being said, I don't mind changing my implementation once more. I just don't see the issue with the current one besides the fact it differs slightly from some other apps/samples.

Reply:
OK I see your point.  It's OK.